src/upgrades/mod.rs

```diff
@@ -220,79 +223,335 @@ impl UpgradeMode {
+                {
+                    self.store_new_command(&scope, &command)?;
+                    data_upgrade_info.increment_command();
+                }
+
+                // Report progress and expected time to finish on every 100 commands evaluated.
+                total_migrated += 1;
+                if total_migrated % 100 == 0 {
+                    // expected time: (total_migrated / (now - started)) * total
+
+                    let mut time_passed = (Time::now().timestamp() - time_started.timestamp()) as usize;
+                    if time_passed == 0 {
+                        time_passed = 1; // avoid divide by zero.. we are doing approximate estimates here
+                    }
+                    let migrated_per_second: f64 = total_migrated as f64 / time_passed as f64;
+                    let expected_seconds = (total_commands as f64 / migrated_per_second) as i64;
+                    let eta = time_started + chrono::Duration::seconds(expected_seconds);
+                    info!(
+                        "  migrated {} commands, expect to finish: {}",
+                        total_migrated,
+                        eta.to_rfc3339()
+                    );
+                }
+            }
+
+            info!("Finished migrating commands for '{}'", scope);
+
+            // Verify migration
+            info!(
+                "Will verify the migration by rebuilding '{}' from migrated commands",
+                &scope
+            );
+            let latest = self.preparation_aggregate_store().get_latest(&handle).map_err(|e| {
+                UpgradeError::Custom(format!(
+                    "Could not rebuild state after migrating CA '{}'! Error was: {}.",
+                    handle, e
+                ))
+            })?;
+
+            // Store snapshot to avoid having to re-process the deltas again in future
+            self.preparation_aggregate_store()
+                .store_snapshot(&handle, latest.as_ref())
+                .map_err(|e| {
+                    UpgradeError::Custom(format!(
+                        "Could not save snapshot for CA '{}' after migration! Disk full?!? Error was: {}.",
+                        handle, e
+                    ))
+                })?;
+
+            // Call the post command migration hook, this will do nothing
+            // unless the implementer of this trait overrode it.
+            self.post_migration(&handle)?;
+
+            // Update the upgrade info as this could be a prepare only
+            // run, and this migration could be resumed later after more
+            // changes were applied.
+            self.update_data_upgrade_info(&scope, &data_upgrade_info)?;
+
+            info!("Verified migration of '{}'", handle);
+        }
+
+        match mode {
+            UpgradeMode::PrepareOnly => {
+                info!(
+                    "Prepared migrating CAs to Krill version {}. Will save progress for final upgrade when Krill restarts.",
```

Comment:
Above you no longer refer to CA but instead this fn has become more generic. Is it correct to still refer to CA here?

Reply:
good catch, changed to data